tools/nimdigger.nim

```diff
@@ -0,0 +1,321 @@
+
+proc main2(opt: DiggerOpt) =
+  let state = DiggerState(nimDir: opt.nimDir, rev: opt.rev)
+  if state.nimDir.len == 0:
+    let nimdiggerHome = getEnv(NimDiggerEnv, getHomeDir() / ".nimdigger")
+    state.nimDir = nimdiggerHome / "cache/Nim"
+  if verbose: dbg state
+  let nimDir = state.nimDir
+  state.binDir = nimDir/"bin"
+
+  if nimDir.dirExists:
+    doAssert fileExists(nimDir / "lib/system.nim"), fmt"nimDir is not a nim repo: {nimDir}"
+  else:
+    createDir nimDir.parentDir
+    gitClone("https://github.com/nim-lang/Nim", nimDir)
+  state.coptv0 = CsourcesOpt(dir: nimDir/"csources", url: "https://github.com/nim-lang/csources.git", name: "csources", revs: csourcesRevs)
+  state.coptv1 = CsourcesOpt(dir: nimDir/"csources_v1", url: "https://github.com/nim-lang/csources_v1.git", name: "csources_v1", revs: csourcesV1Revs)
+  for copt in [state.coptv0, state.coptv1]:
+    copt.binDir = state.binDir
+    copt.fetch = opt.fetch
+    if opt.buildAllCsources:
+      buildCsourcesAnyRevs(copt)
+
+  if opt.fetch: gitFetch(nimDir)
+  if state.rev.len > 0:
+    gitResetHard(nimDir, state.rev)
+  state.rev = gitCurrentRev(state.nimDir)
+  let nimDiggerExe = state.binDir / fmt"nim_nimdigger_nim_{state.rev}{ExeExt2}"
+  if opt.compileNim:
+    if not nimDiggerExe.fileExists:
+      let copt = getNimCsourcesAnyExe(state)
+      buildCsourcesRev(copt)
+      discard runCmdOutput(fmt"{copt.nimCsourcesExe} c -o:{nimDiggerExe} --hints:off --skipUserCfg compiler/nim.nim", nimDir)
+    copyFile(nimDiggerExe, state.binDir / "nim" & ExeExt2)
+
+  if opt.oldnew.len > 0:
+    let oldnew2 = opt.oldnew.split("..")
+    doAssert oldnew2.len == 2, opt.oldnew
+    let oldrev = oldnew2[0]
+    let newrev = oldnew2[1]
+    doAssert oldrev.len > 0 # for regressions, aka goodrev
+    doAssert newrev.len > 0 # for a regressions, aka badrev 
+    gitCleanDanger(state.nimDir, requireConfirmation = true)
+    proc bisectStart(old, new: string)=
+      runCmd(fmt"git -C {state.nimDir.quoteShell} bisect start --term-old {old} --term-new {new} {newrev} {oldrev}")
+    if opt.bisectBugfix: bisectStart("BROKEN", "BUGFIX")
+    else: bisectStart("WORKS", "REGRESSION")
+    let exe = getAppFileName()
+    var msg2 = opt.bisectCmd
+    if opt.bisectBugfix:
+      msg2 = fmt"! ({msg2})" # negate exit code
+    let bisectCmd2 = fmt"{exe} --compileNim && ( {msg2} )"
+    runCmd(fmt"git -C {state.nimDir.quoteShell} bisect run bash -c {bisectCmd2.quoteShell}")
+
+proc main(rev = "", nimDir = "", compileNim = false, fetch = false, oldnew = "", bisectBugfix = false, verbose = false, bisectCmd = "", args: seq[string]) =
+  nimdigger.verbose = verbose
+  var bisectCmd = bisectCmd
+  if bisectCmd.len == 0:
+    bisectCmd = args.quoteShellCommand
+  else:
+    doAssert args.len == 0
+  main2(DiggerOpt.ctor(rev, nimDir, compileNim, fetch, bisectCmd, oldnew, bisectBugfix))
+
+when isMainModule:
+  import pkg/cligen
+  dispatch main
```

Comment:
Maybe would be great to be the first tool to implement the idea of having a `--dump`,
that outputs machine readable useful information as pretty printed JSON, 
then the same design can be implemented on other tools later, see
https://github.com/nim-lang/Nim/issues/13740#issuecomment-603481849

Reply:
yes, I'm all for `--dump:output.json`, but this can be done as followup work, and is not hard with jsonutils, I've added it as future work on top